File: Cores/JVM/src/main/scala/OSOL/Extremum/Cores/JVM/Vectors/VectorObject.scala
package OSOL.Extremum.Cores.JVM.Vectors

import Exceptions._

abstract class VectorObject[Base] (val elements: Map[String, Base]) {

  final def keys: Set[String] = elements.keySet

  final def apply(key: String): Base = {
    try elements(key)
    catch {
      case _: Exception => throw new MissingKeyException(key)
    }
  }

  final def getOrElse(key: String, default: Base): Base = elements.getOrElse(key, default)
  final def apply(key: String, default: Base): Base = this.getOrElse(key, default)

  override def toString: String = elements.map { case (key, value) => s"$key -> $value" }.mkString("\n")

  final def elementWiseOp(that: VectorObject[Base], op: (Base, Base) => Base): Iterable[(String, Base)] = {
    val keys_1 = this.keys
    val keys_2 = that.keys
    if (keys_1 != keys_2) throw new DifferentKeysException(keys_1, keys_2)
    else keys_1.map(k => (k, op(this(k), that(k))))
  }

  final def elementWiseOpImputeMissingKeys(that: VectorObject[Base], op: (Base, Base) => Base, default: Base): Iterable[(String, Base)] = {
    val mergedKeys = this.keys.union(that.keys)
    mergedKeys.map(k => (k, op(this(k, default), that(k, default))))
  }

  def equalsTo(that: VectorObject[Base]): Boolean
  final def ==(that: VectorObject[Base]): Boolean = this.equalsTo(that)

  def add(that: VectorObject[Base]): VectorObject[Base]
  final def +(that: VectorObject[Base]): VectorObject[Base] = this.add(that)

  def addImputeMissingKeys(that: VectorObject[Base]): VectorObject[Base]
  final def ~+(that: VectorObject[Base]): VectorObject[Base] = this.addImputeMissingKeys(that)

  def multiply(coefficient: Double): VectorObject[Base]
  final def *(coefficient: Double): VectorObject[Base] = this.multiply(coefficient)

  final def neg(): VectorObject[Base] = this.multiply(-1)
  final def unary_-(): VectorObject[Base] = this.neg()

  final def subtract(that: VectorObject[Base]): VectorObject[Base] = this.add(that.multiply(-1))
  final def -(that: VectorObject[Base]): VectorObject[Base] = this.subtract(that)

  final def subtractImputeMissingKeys(that: VectorObject[Base]): VectorObject[Base] = this.addImputeMissingKeys(that.neg())
  final def ~-(that: VectorObject[Base]): VectorObject[Base] = this.subtractImputeMissingKeys(that)

  def multiply(that: VectorObject[Base]): VectorObject[Base]
  final def *(that: VectorObject[Base]): VectorObject[Base] = this.multiply(that)

  def multiplyImputeMissingKeys(that: VectorObject[Base]): VectorObject[Base]
  final def ~*(that: VectorObject[Base]): VectorObject[Base] = this.multiplyImputeMissingKeys(that)

  def union(that: (String, Base)): VectorObject[Base]
  def union(vectors: (String, Base)*): VectorObject[Base] = vectors.foldLeft(this) { case (a, b) => a.union(b) }

  def distanceFromArea(area: Map[String, (Double, Double)]): Map[String, Double]

}

File: Cores/DotNet/Vectors/VectorObject.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace OSOL.Extremum.Cores.DotNet.Vectors
{
    public abstract class VectorObject<TBase>
    {
        public Dictionary<String, TBase> Elements;

        public IEnumerable<string> Keys => this.Elements.Keys;

        public TBase this[string key]
        {
            get
            {
                try
                {
                    return this.Elements[key];
                }
                catch (Exception e)
                {
                    throw new VectorExceptions.MissingKeyException(key);
                }
            }
            set => Elements[key] = value;
        }

        public TBase this[string key, TBase defaultValue] => Keys.Contains(key) ? this[key] : defaultValue;

        public override string ToString() =>
            string.Join(separator: "\n", values: Elements.Select(x => $"{x.Key} -> {x.Value.ToString()}"));

        public Dictionary<String, TBase> ElementWiseOp(VectorObject<TBase> that, Func<TBase, TBase, TBase> op)
        {
            string[] keys_1 = this.Keys.ToArray();
            string[] keys_2 = that.Keys.ToArray();
            if (!(keys_1.All(k => keys_2.Contains(k)) && keys_2.All(k => keys_1.Contains(k))))
            {
                throw new VectorExceptions.DifferentKeysException(keys_1, keys_2);
            }
            else
            {
                return keys_1.ToDictionary(k => k, k => op(this[k], that[k]));
            }
        }

        public Dictionary<String, TBase> ElementWiseOpImputeMissingKeys(VectorObject<TBase> that,
            Func<TBase, TBase, TBase> op, TBase defaultValue)
        {
            List<string> mergedKeys = new List<string>(this.Keys);
            foreach (string k in that.Keys)
            {
                if (!mergedKeys.Contains(k))
                {
                    mergedKeys.Add(k);
                }
            }

            return mergedKeys.ToDictionary(k => k, k => op(this[k, defaultValue], that[k, defaultValue]));
        }

        public abstract bool EqualsTo(VectorObject<TBase> that);
        public static bool operator ==(VectorObject<TBase> a, VectorObject<TBase> b) => a.EqualsTo(b);
        public static bool operator !=(VectorObject<TBase> a, VectorObject<TBase> b) => !(a == b);

        public abstract VectorObject<TBase> Add(VectorObject<TBase> that);
        public static VectorObject<TBase> operator +(VectorObject<TBase> a, VectorObject<TBase> b) => a.Add(b);

        public abstract VectorObject<TBase> AddImputeMissingKeys(VectorObject<TBase> that);

        public abstract VectorObject<TBase> Multiply(double coefficient);

        public static VectorObject<TBase> operator *(VectorObject<TBase> a, double coefficient) =>
            a.Multiply(coefficient);

        private VectorObject<TBase> Neg() => Multiply(coefficient: -1.0);
        public static VectorObject<TBase> operator -(VectorObject<TBase> a) => a.Neg();

        public abstract VectorObject<TBase> Subtract(VectorObject<TBase> that);

        public static VectorObject<TBase> operator -(VectorObject<TBase> a, VectorObject<TBase> b) =>
            a.Subtract(b);

        public abstract VectorObject<TBase> SubtractImputeMissingKeys(VectorObject<TBase> that);

        public abstract VectorObject<TBase> Multiply(VectorObject<TBase> that);

        public static VectorObject<TBase> operator *(VectorObject<TBase> a, VectorObject<TBase> b) =>
            a.Multiply(b);

        public abstract VectorObject<TBase> MultiplyImputeMissingKeys(VectorObject<TBase> that);

        public abstract VectorObject<TBase> Union(params Tuple<string, TBase>[] vectors);

    }
}